File: Blind75/java/array/MaxSubarraySum.java
package Blind75.java.array; 

public class MaxSubarraySum { 

    public int maxSumSubArray(int[] nums) { 
        int max = nums[0]; 
        int currentSum = 0; 

        for(int i = 0; i < nums.length; i++) { 
            if(currentSum < 0) 
                currentSum = 0; 
            currentSum += nums[i]; 
            max = Math.max(currentSum, max);
        }
        return max;
    }


    public int maxSumSubArray2(int[] nums) { 
        int[] dp = new int[nums.length]; 
        dp[0] = nums[0]; 
        int max = dp[0]; 
        
        for (int i = 0; i < nums.length; i++) { 
            dp[i] = nums[i] + (dp[i-1] > 0 ? dp[i-1] : 0);
        } 
        return max;
    }
  
}

File: Blind75/ruby/array/max_subarray_sum.rb
def max_subarray_sum(nums) 
    max_so_far = nums[0] 
    current_sum = 0 
    len = nums.length - 1 
    # [-2,1,-3,4,-1,2,1,-5,4]
    for i in 0..len 
        if current_sum < 0 
            current_sum = 0 
        end 
        current_sum += nums[i]  
        max_so_far = [max_so_far, current_sum].max 
    end  